File: src/howtoprogram/ch1/Tiler.java
package howtoprogram.ch1;

import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;

public class Tiler {

    private int pictureSize = 512;
    private int[][] tileColors;

    public Tiler(int[][] tileColors) {
        this.tileColors = tileColors;
    }

    public Canvas setUpPicture() {
        Canvas canvas = new Canvas(pictureSize, pictureSize);
        //todo warning if not square
        //todo warning if side length not power of 2
        int numberOfTilesPerSide = tileColors.length;
        int sizeOfEachTile = pictureSize / numberOfTilesPerSide;
        GraphicsContext graphics = canvas.getGraphicsContext2D();//Does the drawing.
        for (int row=0; row<numberOfTilesPerSide; row++) {
            for (int column=0; column<numberOfTilesPerSide; column++) {
                int offsetX = column * sizeOfEachTile;
                int offsetY = row * sizeOfEachTile;
                int intensity = tileColors[row][column];
                Color tileColor = Color.rgb(intensity, intensity, intensity);
                graphics.setFill(tileColor);
                graphics.fillRect(offsetX, offsetY, sizeOfEachTile, sizeOfEachTile);
            }
        }
        return canvas;
    }
}

File: src/pfb/basics/Tiler.kt
package pfb.basics

import javafx.scene.canvas.Canvas
import javafx.scene.canvas.GraphicsContext
import javafx.scene.paint.Color

class Tiler(private val tileColors: Array<Array<Int>>) {

    private val pictureSize = 200

    fun setUpPicture(): Canvas {
        val canvas = Canvas(pictureSize.toDouble(), pictureSize.toDouble())
        val numberOfTilesPerSide = tileColors.size
        val sizeOfEachTile = pictureSize / numberOfTilesPerSide
        val graphics = canvas.graphicsContext2D//Does the drawing.
        for (row in 0..numberOfTilesPerSide - 1) {
            for (column in 0..numberOfTilesPerSide - 1) {
                val offsetX = column * sizeOfEachTile
                val offsetY = row * sizeOfEachTile
                val intensity = tileColors[row][column]
                val tileColor = Color.rgb(intensity, intensity, intensity)
                graphics.fill = tileColor
                graphics.fillRect(offsetX.toDouble(), offsetY.toDouble(), sizeOfEachTile.toDouble(), sizeOfEachTile.toDouble())
            }
        }
        return canvas
    }
}